UVerb HasServer Instance

Been trying to debug this all day but I think it's not working because maybe there's an interface missing for servant-auth with UVerbs.  Maybe that's not supported yet?

The error I'm getting is as follows.
```
/home/robwithhair/GGMR/vfde-inv-var-api/src/API.hs:112:55: error:
    • No instance for (HasServer  
                         (Servant.Auth.Server.Internal.AddSetCookie.AddSetCookieApi
                            (Servant.Auth.Server.Internal.AddSetCookie.AddSetCookieApi
                               (UVerb
                                  'PUT
                                  '[JSON]
                                  '[WithStatus 201 NoContent, WithStatus 204 NoContent])))
                         '[JWTSettings, CookieSettings,
                           BasicAuthData -> IO (AuthResult AuthenticatedUser)])
        arising from a use of ‘serveWithContext’
    • In the second argument of ‘($)’, namely
        ‘serveWithContext api cfg (server config)’
      In a stmt of a 'do' block:  
        pure . ifRequest shouldApplyApiGateway apiGateway
          $ serveWithContext api cfg (server config)
      In the expression:          
        do myKey <- fromSecret <$> ByteString.readFile jwtSecretPath
           let jwtCfg = defaultJWTSettings myKey
               authCfg = authCheck config
               ....               
           pure . ifRequest shouldApplyApiGateway apiGateway
             $ serveWithContext api cfg (server config)
    |                             
112 |   pure . ifRequest shouldApplyApiGateway apiGateway $ serveWithContext api cfg (server config)
    |                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                  
```

Not sure what I'm missing here, so would be infinitely grateful for any input.

```
type LocallyStoredItemAPI (sectionName :: Symbol) listType detailType patchType patchResponseType = sectionName :>
                                            ( Get '[JSON] listType    -- LIST
                                              :<|> Capture "id" ItemID :> Get '[JSON] detailType  -- GET
                                              :<|> Capture "id" ItemID :> ReqBody '[JSON] detailType :> UVerb 'PUT '[JSON] '[WithStatus 201 NoContent, WithStatus 204 NoContent]  -- PUT
                                              :<|> Capture "id" ItemID :> ReqBody '[JSON] patchType :> Patch '[JSON] patchResponseType -- PATCH
                                              :<|> ReqBody '[JSON] detailType :> PostCreated '[JSON] NoContent -- POST
                                              :<|> Capture "id" ItemID :> Delete '[JSON] NoContent -- DELETE
                                           )

putLocallyStoredItem :: (SetModifiedTime i, ToJSON i)
  => Config
  -> (LocallyStoredItemId -> C.RelativePath)
  -> LocallyStoredItemId
  -> i
  -> Handler (Union '[WithStatus 201 NoContent, WithStatus 204 NoContent])
putLocallyStoredItem cfg dsJsonFileFromId dsid body = do
  itemExistsBefore <- liftIO $ (C.fileExistsF cfg) (dsJsonFileFromId dsid)
  -- Write to disk with modified time
  liftIO $ withCurrentModifiedTime body >>= writeDsFile . encode
  if itemExistsBefore then
    Servant.Server.UVerb.respond (WithStatus @204 NoContent)
  else
    Servant.Server.UVerb.respond (WithStatus @201 NoContent)
  where writeDsFile = C.writeFileF cfg (dsJsonFileFromId dsid)
```
I realise this isn't a very minimal example.  If I can I'll work on one but on a deadline ATM.  Hope you understand.  I'm following the examples from servant Cookbook.  Looked everywhere for an import to add the interfaces but think writing them myself may be beyond my understanding of Servant ATM.

@and-pete from not enough interfaces to too many!  I'm now getting the following and I have tried compiling with IncoherentInstances too but no change from below error.  Running gdeest's pull request now.

```
/home/robwithhair/GGMR/vfde-inv-var-api/src/API.hs:94:55: error:
    • Overlapping instances for HasServer
                                  (Servant.Auth.Server.Internal.AddSetCookie.AddSetCookieApi
                                     (Servant.Auth.Server.Internal.AddSetCookie.AddSetCookieApi
                                        (UVerb
                                           'PUT
                                           '[JSON]
                                           '[WithStatus 201 NoContent, WithStatus 204 NoContent])))
                                  '[JWTSettings, CookieSettings,
                                    BasicAuthData -> IO (AuthResult AuthenticatedUser)]
        arising from a use of ‘serveWithContext’
      Matching instances:
        instance [overlappable] forall k (api :: k) (context :: [*]).
                                (TypeError ...) =>
                                HasServer api context
          -- Defined in ‘Servant.Server.Internal’
        ...plus 32 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
      (The choice depends on the instantiation of ‘’
       To pick the first instance above, use IncoherentInstances
       when compiling the other instance declarations)
    • In the second argument of ‘($)’, namely
        ‘serveWithContext api cfg (server config)’
      In a stmt of a 'do' block:
        pure . ifRequest shouldApplyApiGateway apiGateway
          $ serveWithContext api cfg (server config)
      In the expression:
        do myKey <- fromSecret <$> ByteString.readFile jwtSecretPath
           let jwtCfg = defaultJWTSettings myKey
               authCfg = authCheck config
               ....
           pure . ifRequest shouldApplyApiGateway apiGateway
             $ serveWithContext api cfg (server config)
   |
94 |   pure . ifRequest shouldApplyApiGateway apiGateway $ serveWithContext api cfg (server config)
   |   
```

@gdeest I'm getting the following...

```
/home/robwithhair/GGMR/vfde-inv-var-api/src/API.hs:94:55: error:
    • Duplicate element in list:  
          '[StatusOf              
              (Headers            
                 '[Header "Set-Cookie" SetCookie, Header "Set-Cookie" SetCookie]
                 (WithStatus 201 NoContent)),
            StatusOf              
              (Headers            
                 '[Header "Set-Cookie" SetCookie, Header "Set-Cookie" SetCookie]
                 (WithStatus 204 NoContent))]
    • In the second argument of ‘($)’, namely
        ‘serveWithContext api cfg (server config)’
      In a stmt of a 'do' block:  
        pure . ifRequest shouldApplyApiGateway apiGateway
          $ serveWithContext api cfg (server config)
      In the expression:          
        do myKey <- fromSecret <$> ByteString.readFile jwtSecretPath
           let jwtCfg = defaultJWTSettings myKey
               authCfg = authCheck config
               ....               
           pure . ifRequest shouldApplyApiGateway apiGateway
             $ serveWithContext api cfg (server config)
   |                              
94 |   pure . ifRequest shouldApplyApiGateway apiGateway $ serveWithContext api cfg (server config)
   |                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                  
Completed 242 action(s).
```

@gdeest building from 9030ea442cd4acaef74e8d5c9a6aa70f392a8d89 gives me the following

```
/home/robwithhair/GGMR/vfde-inv-var-api/src/API.hs:94:55: error:
    • Duplicate element in list:
          '[StatusOf
              (Headers
                 '[Header "Set-Cookie" SetCookie, Header "Set-Cookie" SetCookie]
                 (WithStatus 201 NoContent)),
            StatusOf
              (Headers
                 '[Header "Set-Cookie" SetCookie, Header "Set-Cookie" SetCookie]
                 (WithStatus 204 NoContent))]
    • In the second argument of ‘($)’, namely
        ‘serveWithContext api cfg (server config)’
      In a stmt of a 'do' block:
        pure . ifRequest shouldApplyApiGateway apiGateway
          $ serveWithContext api cfg (server config)
      In the expression:
        do myKey <- fromSecret <$> ByteString.readFile jwtSecretPath
           let jwtCfg = defaultJWTSettings myKey
               authCfg = authCheck config
               ....
           pure . ifRequest shouldApplyApiGateway apiGateway
             $ serveWithContext api cfg (server config)
   |
94 |   pure . ifRequest shouldApplyApiGateway apiGateway $ serveWithContext api cfg (server config)
   |                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


```

@gdeest similar error again clean build of 6bf95544050eb0333e6b115b048274be032a05b8
```
/home/robwithhair/GGMR/vfde-inv-var-api/src/API.hs:94:55: error:
    • Duplicate element in list:
          '[StatusOf
              (Headers
                 '[Header "Set-Cookie" SetCookie, Header "Set-Cookie" SetCookie]
                 (WithStatus 201 NoContent)),
            StatusOf
              (Headers
                 '[Header "Set-Cookie" SetCookie, Header "Set-Cookie" SetCookie]
                 (WithStatus 204 NoContent))]
    • In the second argument of ‘($)’, namely
        ‘serveWithContext api cfg (server config)’
      In a stmt of a 'do' block:
        pure . ifRequest shouldApplyApiGateway apiGateway
          $ serveWithContext api cfg (server config)
      In the expression:
        do myKey <- fromSecret <$> ByteString.readFile jwtSecretPath
           let jwtCfg = defaultJWTSettings myKey
               authCfg = authCheck config
               ....
           pure . ifRequest shouldApplyApiGateway apiGateway
             $ serveWithContext api cfg (server config)
   |
94 |   pure . ifRequest shouldApplyApiGateway apiGateway $ serveWithContext api cfg (server config)
   |                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

@gdeest new error now as follows when clean building 0dd77adec87d25940f40e5342af09ea88ea819f2

```
/home/robwithhair/GGMR/vfde-inv-var-api/src/API.hs:94:55: error:
    • Couldn't match type ‘sop-core-0.5.0.2:Data.SOP.NS.NS
                             sop-core-0.5.0.2:Data.SOP.BasicFunctors.I
                             '[Headers
                                 '[Header "Set-Cookie" SetCookie, Header "Set-Cookie" SetCookie]
                                 (WithStatus 201 NoContent),
                               Headers
                                 '[Header "Set-Cookie" SetCookie, Header "Set-Cookie" SetCookie]
                                 (WithStatus 204 NoContent)]’
                     with ‘Headers '[Header "Set-Cookie" SetCookie] cookied1’
        arising from a use of ‘serveWithContext’
    • In the second argument of ‘($)’, namely
        ‘serveWithContext api cfg (server config)’
      In a stmt of a 'do' block:
        pure . ifRequest shouldApplyApiGateway apiGateway
          $ serveWithContext api cfg (server config)
      In the expression:
        do myKey <- fromSecret <$> ByteString.readFile jwtSecretPath
           let jwtCfg = defaultJWTSettings myKey
               authCfg = authCheck config
               ....
           pure . ifRequest shouldApplyApiGateway apiGateway
             $ serveWithContext api cfg (server config)
   |
94 |   pure . ifRequest shouldApplyApiGateway apiGateway $ serveWithContext api cfg (server config)
   |                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

--- a/changelog.d/1571
+++ b/changelog.d/1571
@@ -1,0 +1,8 @@
+synopsis: Support UVerb in servant-auth-server
+prs: #1571
+issues: #1570
+description: {
+UVerb endpoints are now supported by servant-auth-server and can used under the
+Auth combinator when writing servers. It is still unsupported by
+servant-auth-client.
+}

--- a/changelog.d/1571
+++ b/changelog.d/1571
@@ -2,7 +2,7 @@ synopsis: Support UVerb in servant-auth-server
 prs: #1571
 issues: #1570
 description: {
-UVerb endpoints are now supported by servant-auth-server and can used under the
+UVerb endpoints are now supported by servant-auth-server and can be used under the
 Auth combinator when writing servers. It is still unsupported by
 servant-auth-client.
 }

--- a/servant-auth/servant-auth-server/servant-auth-server.cabal
+++ b/servant-auth/servant-auth-server/servant-auth-server.cabal
@@ -129,6 +129,7 @@ test-suite spec
     , lens-aeson  >= 1.0.2     && < 1.2
     , warp        >= 3.2.25    && < 3.4
     , wreq        >= 0.5.2.1   && < 0.6
+    , text        >= 1.2.3.0   && < 2.1
   other-modules:
       Servant.Auth.ServerSpec
   default-language: Haskell2010

--- a/servant-auth/servant-auth-server/src/Servant/Auth/Server/Internal.hs
+++ b/servant-auth/servant-auth-server/src/Servant/Auth/Server/Internal.hs
@@ -54,15 +54,12 @@ instance ( n ~ 'S ('S 'Z)
 
       makeCookies :: AuthResult v -> IO (SetCookieList ('S ('S 'Z)))
       makeCookies authResult = do
-        xsrf <- makeXsrfCookie cookieSettings
-        fmap (Just xsrf `SetCookieCons`) $
-          case authResult of
-            (Authenticated v) -> do
-              ejwt <- makeSessionCookie cookieSettings jwtSettings v
-              case ejwt of
-                Nothing  -> return $ Nothing `SetCookieCons` SetCookieNil
-                Just jwt -> return $ Just jwt `SetCookieCons` SetCookieNil
-            _ -> return $ Nothing `SetCookieCons` SetCookieNil
+        case authResult of
+          (Authenticated v) -> do
+            ejwt <- makeSessionCookie cookieSettings jwtSettings v
+            xsrf <- makeXsrfCookie cookieSettings
+            return $ Just xsrf `SetCookieCons` (ejwt `SetCookieCons` SetCookieNil)
+          _ -> return $ Nothing `SetCookieCons` (Nothing `SetCookieCons` SetCookieNil)
 
       go :: (AuthResult v -> ServerT api Handler)
          -> (AuthResult v, SetCookieList n)

--- a/servant-auth/servant-auth-server/src/Servant/Auth/Server/Internal/AddSetCookie.hs
+++ b/servant-auth/servant-auth-server/src/Servant/Auth/Server/Internal/AddSetCookie.hs
@@ -11,6 +11,7 @@ import           Data.Tagged              (Tagged (..))
 import qualified Network.HTTP.Types       as HTTP
 import           Network.Wai              (mapResponseHeaders)
 import           Servant
+import           Servant.API.UVerb.Union
 import           Servant.API.Generic
 import           Servant.Server.Generic
 import           Web.Cookie
@@ -33,12 +34,24 @@ type family AddSetCookieApiVerb a where
   AddSetCookieApiVerb (Headers ls a) = Headers (Header "Set-Cookie" SetCookie ': ls) a
   AddSetCookieApiVerb a = Headers '[Header "Set-Cookie" SetCookie] a
 
+#if MIN_VERSION_servant_server(0,18,1)
+type family MapAddSetCookieApiVerb (as :: [*]) where
+  MapAddSetCookieApiVerb '[] = '[]
+  MapAddSetCookieApiVerb (a ': as) = (AddSetCookieApiVerb a ': MapAddSetCookieApiVerb as)
+#endif
+
 type family AddSetCookieApi a :: *
 type instance AddSetCookieApi (a :> b) = a :> AddSetCookieApi b
 type instance AddSetCookieApi (a :<|> b) = AddSetCookieApi a :<|> AddSetCookieApi b
+#if MIN_VERSION_servant_server(0,19,0)
 type instance AddSetCookieApi (NamedRoutes api) = AddSetCookieApi (ToServantApi api)
+#endif
 type instance AddSetCookieApi (Verb method stat ctyps a)
   = Verb method stat ctyps (AddSetCookieApiVerb a)
+#if MIN_VERSION_servant_server(0,18,1)
+type instance AddSetCookieApi (UVerb method ctyps as)
+  = UVerb method ctyps (MapAddSetCookieApiVerb as)
+#endif
 type instance AddSetCookieApi Raw = Raw
 #if MIN_VERSION_servant_server(0,15,0)
 type instance AddSetCookieApi (Stream method stat framing ctyps a)
@@ -57,7 +70,7 @@ instance {-# OVERLAPS #-} AddSetCookies ('S n) oldb newb
   => AddSetCookies ('S n) (a -> oldb) (a -> newb) where
   addSetCookies cookies oldfn = addSetCookies cookies . oldfn
 
-instance AddSetCookies 'Z orig orig where
+instance (orig1 ~ orig2) => AddSetCookies 'Z orig1 orig2 where
   addSetCookies _ = id
 
 instance {-# OVERLAPPABLE #-}

--- a/servant-auth/servant-auth-server/test/Servant/Auth/ServerSpec.hs
+++ b/servant-auth/servant-auth-server/test/Servant/Auth/ServerSpec.hs
@@ -26,7 +26,7 @@ import           Data.Aeson                          (FromJSON, ToJSON, Value,
 import           Data.Aeson.Lens                     (_JSON)
 import qualified Data.ByteString                     as BS
 import qualified Data.ByteString.Lazy                as BSL
-import           Data.Text                           (Text)
+import           Data.Text                           (Text, pack)
 import           Data.CaseInsensitive                (mk)
 import           Data.Foldable                       (find)
 import           Data.Monoid
@@ -42,6 +42,7 @@ import           Network.HTTP.Types                  (Status, status200,
 import           Network.Wai                         (responseLBS)
 import           Network.Wai.Handler.Warp            (testWithApplication)
 import           Network.Wreq                        (Options, auth, basicAuth,
+                                                      checkResponse,
                                                       cookieExpiryTime, cookies,
                                                       defaults, get, getWith, postWith,
                                                       header, oauth2Bearer,
@@ -185,8 +186,21 @@ cookieAuthSpec
         it "fails with no XSRF header or cookie" $ \port -> property
                                                          $ \(user :: User) -> do
           jwt <- createJWT theKey (newJWSHeader ((), HS256)) (claims $ toJSON user)
-          opts <- addJwtToCookie cookieCfg jwt
-          getWith opts (url port) `shouldHTTPErrorWith` status401
+          opts' <- addJwtToCookie cookieCfg jwt
+          let opts = opts' & checkResponse .~ Just mempty
+          resp <- getWith opts (url port)
+          resp ^. responseStatus `shouldBe` status401
+          (resp ^. responseCookieJar) `shouldNotHaveCookies` ["XSRF-TOKEN"]
+
+          -- Validating that the XSRF cookie isn't added for UVerb routes either.
+          -- These routes can return a 401 response directly without using throwError / throwAll,
+          -- which revealed a bug:
+          --
+          -- https://github.com/haskell-servant/servant/issues/1570#issuecomment-1076374449
+          resp <- getWith opts (url port ++ "/uverb")
+          resp ^. responseStatus `shouldBe` status401
+          (resp ^. responseCookieJar) `shouldNotHaveCookies` ["XSRF-TOKEN"]
+
 
         it "succeeds if XSRF header and cookie match, and JWT is valid" $ \port -> property
                                                                         $ \(user :: User) -> do
@@ -408,14 +422,14 @@ type API auths
     = Auth auths User :>
         ( Get '[JSON] Int
        :<|> ReqBody '[JSON] Int :> Post '[JSON] Int
-       :<|> NamedRoutes DummyRoutes
-       :<|> UVerb 'GET '[JSON] '[WithStatus 200 Int, WithStatus 500 Text]
+       :<|> "named" :> NamedRoutes DummyRoutes
        :<|> "header" :> Get '[JSON] (Headers '[Header "Blah" Int] Int)
 #if MIN_VERSION_servant_server(0,15,0)
        :<|> "stream" :> StreamGet NoFraming OctetStream (SourceIO BS.ByteString)
 #endif
        :<|> "raw" :> Raw
         )
+      :<|> "uverb" :> Auth auths User :> UVerb 'GET '[JSON] '[WithStatus 200 Int, WithStatus 401 Text, WithStatus 403 Text]
       :<|> "login" :> ReqBody '[JSON] User :> Post '[JSON] (Headers '[ Header "Set-Cookie" SetCookie
                                                                      , Header "Set-Cookie" SetCookie ] NoContent)
       :<|> "logout" :> Get '[JSON] (Headers '[ Header "Set-Cookie" SetCookie
@@ -486,7 +500,6 @@ server ccfg =
         Authenticated usr -> getInt usr
                         :<|> postInt usr
                         :<|> DummyRoutes { dummyInt = getInt usr }
-                        :<|> respond (WithStatus @200 (42 :: Int))
                         :<|> getHeaderInt
 #if MIN_VERSION_servant_server(0,15,0)
                         :<|> return (S.source ["bytestring"])
@@ -494,6 +507,11 @@ server ccfg =
                         :<|> raw
         Indefinite -> throwAll err401
         _ -> throwAll err403
+    ) :<|>
+    (\authResult -> case authResult of
+      Authenticated usr -> respond (WithStatus @200 (42 :: Int))
+      Indefinite -> respond (WithStatus @401 $ pack "Authentication required")
+      _ -> respond (WithStatus @403 $ pack "Forbidden")
     )
     :<|> getLogin
     :<|> getLogout
@@ -573,6 +591,12 @@ shouldMatchCookieNames :: HCli.CookieJar -> [BS.ByteString] -> Expectation
 shouldMatchCookieNames cj patterns
     = fmap cookie_name (destroyCookieJar cj)
     `shouldMatchList` patterns
+
+shouldNotHaveCookies :: HCli.CookieJar -> [BS.ByteString] -> Expectation
+shouldNotHaveCookies cj patterns =
+    sequence_ $ (\cookieName -> cookieNames `shouldNotContain` [cookieName]) <$> patterns
+  where cookieNames :: [BS.ByteString]
+        cookieNames = cookie_name <$> destroyCookieJar cj
 
 shouldMatchCookieNameValues :: HCli.CookieJar -> [(BS.ByteString, BS.ByteString)] -> Expectation
 shouldMatchCookieNameValues cj patterns

--- a/servant-auth/servant-auth-server/test/Servant/Auth/ServerSpec.hs
+++ b/servant-auth/servant-auth-server/test/Servant/Auth/ServerSpec.hs
@@ -1,4 +1,5 @@
 {-# LANGUAGE CPP #-}
+{-# LANGUAGE TypeApplications #-}
 module Servant.Auth.ServerSpec (spec) where
 
 #if !MIN_VERSION_servant_server(0,16,0)
@@ -25,6 +26,7 @@ import           Data.Aeson                          (FromJSON, ToJSON, Value,
 import           Data.Aeson.Lens                     (_JSON)
 import qualified Data.ByteString                     as BS
 import qualified Data.ByteString.Lazy                as BSL
+import           Data.Text                           (Text)
 import           Data.CaseInsensitive                (mk)
 import           Data.Foldable                       (find)
 import           Data.Monoid
@@ -407,6 +409,7 @@ type API auths
         ( Get '[JSON] Int
        :<|> ReqBody '[JSON] Int :> Post '[JSON] Int
        :<|> NamedRoutes DummyRoutes
+       :<|> UVerb 'GET '[JSON] '[WithStatus 200 Int, WithStatus 500 Text]
        :<|> "header" :> Get '[JSON] (Headers '[Header "Blah" Int] Int)
 #if MIN_VERSION_servant_server(0,15,0)
        :<|> "stream" :> StreamGet NoFraming OctetStream (SourceIO BS.ByteString)
@@ -483,6 +486,7 @@ server ccfg =
         Authenticated usr -> getInt usr
                         :<|> postInt usr
                         :<|> DummyRoutes { dummyInt = getInt usr }
+                        :<|> respond (WithStatus @200 (42 :: Int))
                         :<|> getHeaderInt
 #if MIN_VERSION_servant_server(0,15,0)
                         :<|> return (S.source ["bytestring"])

--- a/servant/src/Servant/API/ResponseHeaders.hs
+++ b/servant/src/Servant/API/ResponseHeaders.hs
@@ -8,6 +8,7 @@
 {-# LANGUAGE PolyKinds              #-}
 {-# LANGUAGE ScopedTypeVariables    #-}
 {-# LANGUAGE TypeFamilies           #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
 {-# LANGUAGE TypeOperators          #-}
 {-# LANGUAGE UndecidableInstances   #-}
 {-# OPTIONS_HADDOCK not-home        #-}
@@ -51,6 +52,9 @@ import           Prelude ()
 import           Prelude.Compat
 import           Servant.API.Header
                  (Header)
+import           Servant.API.UVerb.Union
+import qualified Data.SOP.BasicFunctors as SOP
+import qualified Data.SOP.NS as SOP
 
 -- | Response Header objects. You should never need to construct one directly.
 -- Instead, use 'addOptionalHeader'.
@@ -169,6 +173,21 @@ instance {-# OVERLAPPING #-} ( KnownSymbol h, ToHttpApiData v )
 instance {-# OVERLAPPABLE #-} ( KnownSymbol h, ToHttpApiData v , new ~ Headers '[Header h v] a)
          => AddHeader h v a new where
     addOptionalHeader hdr resp = Headers resp (HCons hdr HNil)
+
+-- Instances to decorate all responses in a 'Union' with headers. The functional
+-- dependencies force us to consider singleton lists as the base case in the
+-- recursion (it is impossible to determine h and v otherwise from old / new
+-- responses if the list is empty).
+instance (AddHeader h v old new) => AddHeader h v (Union '[old]) (Union '[new]) where
+  addOptionalHeader hdr resp =
+    SOP.Z $ SOP.I $ addOptionalHeader hdr $ SOP.unI $ SOP.unZ $ resp
+
+instance
+  (AddHeader h v old new, AddHeader h v (Union oldrest) (Union newrest), oldrest ~ (a ': as), newrest ~ (b ': bs))
+  => AddHeader h v (Union (old ': (a ': as))) (Union (new ': (b ': bs))) where
+  addOptionalHeader hdr resp = case resp of
+    SOP.Z (SOP.I rHead) -> SOP.Z $ SOP.I $ addOptionalHeader hdr rHead
+    SOP.S rOthers -> SOP.S $ addOptionalHeader hdr rOthers
 
 -- | @addHeader@ adds a header to a response. Note that it changes the type of
 -- the value in the following ways:

--- a/servant/src/Servant/API/ResponseHeaders.hs
+++ b/servant/src/Servant/API/ResponseHeaders.hs
@@ -183,7 +183,11 @@ instance (AddHeader h v old new) => AddHeader h v (Union '[old]) (Union '[new]) 
     SOP.Z $ SOP.I $ addOptionalHeader hdr $ SOP.unI $ SOP.unZ $ resp
 
 instance
-  (AddHeader h v old new, AddHeader h v (Union oldrest) (Union newrest), oldrest ~ (a ': as), newrest ~ (b ': bs))
+  ( AddHeader h v old new, AddHeader h v (Union oldrest) (Union newrest)
+  -- This ensures that the remainder of the response list is _not_ empty
+  -- It is necessary to prevent the two instances for union types from
+  -- overlapping.
+  , oldrest ~ (a ': as), newrest ~ (b ': bs))
   => AddHeader h v (Union (old ': (a ': as))) (Union (new ': (b ': bs))) where
   addOptionalHeader hdr resp = case resp of
     SOP.Z (SOP.I rHead) -> SOP.Z $ SOP.I $ addOptionalHeader hdr rHead

--- a/servant/src/Servant/API/UVerb.hs
+++ b/servant/src/Servant/API/UVerb.hs
@@ -38,6 +38,7 @@ import GHC.TypeLits (Nat)
 import Network.HTTP.Types (Status, StdMethod)
 import Servant.API.ContentTypes (JSON, PlainText, FormUrlEncoded, OctetStream, NoContent, MimeRender(mimeRender), MimeUnrender(mimeUnrender))
 import Servant.API.Status (KnownStatus, statusVal)
+import Servant.API.ResponseHeaders (Headers)
 import Servant.API.UVerb.Union
 
 class KnownStatus (StatusOf a) => HasStatus (a :: *) where
@@ -86,6 +87,8 @@ newtype WithStatus (k :: Nat) a = WithStatus a
 instance KnownStatus n => HasStatus (WithStatus n a) where
   type StatusOf (WithStatus n a) = n
 
+instance HasStatus a => HasStatus (Headers ls a) where
+  type StatusOf (Headers ls a) = StatusOf a
 
 -- | A variant of 'Verb' that can have any of a number of response values and status codes.
 --

